### PR TITLE
[267] Rule exec in genquery

### PR DESCRIPTION
**Pull request by d-w-moore**

### CODE REVIEW

**Comment by d-w-moore** [comment]
Still need to write the test....
Sample use is below.

**Comment by d-w-moore** [comment]
```python
#!/usr/bin/env python
from __future__ import print_function
import datetime as dt
from  datetime import datetime
import atexit, os, ssl, sys, pprint
import irods
from irods.session import iRODSSession
from irods.query import Query
from irods.models import RuleExec
from irods.rule import Rule

IRODS_SESSION_AUTO_CLEANUP = lambda s:(lambda *_:s.cleanup())

def session_(exit_handler= IRODS_SESSION_AUTO_CLEANUP):
    try:
        env_file = os.environ['IRODS_ENVIRONMENT_FILE']
    except KeyError:
        env_file = os.path.expanduser('~/.irods/irods_environment.json')
    ssl_context = ssl.create_default_context( purpose=ssl.Purpose.SERVER_AUTH,
                                              cafile=None, capath=None, cadata=None)
    ssl_settings = {'ssl_context': ssl_context}
    session = iRODSSession(irods_env_file=env_file, **ssl_settings)
    if exit_handler:
        atexit.register(exit_handler(session))
    return session

if __name__ == '__main__':

    ses = session_(exit_handler = None)

    # rule id's to delete as program args

    if sys.argv[1:] :
       irods.rule.Rule(ses).remove_by_id(*sys.argv[1:])
       exit()

    # if no args, do a sample query
    
    res = [r for r in ses.query(RuleExec).filter(RuleExec.name != '') ]

    # print out exe times: ( + is future, - is past )

    def relative_to_now(t):
        if t is None: return 'None'
        now = datetime.utcnow()
        if t < now:
            return "-{}".format(now-t)
        else:
            return "+{}".format(t-now)

    if res:
        print(relative_to_now (res[0][RuleExec.last_exe_time]))
        print(relative_to_now (res[0][RuleExec.exec_time]))
        print('result from rule query : ')
        pprint.pprint (res[0])
        print('rule owner is',(res[0])[RuleExec.user_name])
        ses.cleanup()
```

File: irods/models.py [hunk 1]
```diff
@@ -22,22 +22,19 @@ class Model(six.with_metaclass(ModelBase, object)):
...
-#('RULE_EXEC_ESTIMATED_EXE_TIME', 1008)
-#('RULE_EXEC_NOTIFICATION_ADDR', 1009)
-#(DateTime,'RULE_EXEC_LAST_EXE_TIME', 1010)
-#('RULE_EXEC_STATUS', 1011)
-#('RULE_EXEC_CONTEXT', 1012)
+# - Unused, or usually absent:
+
+    # exec_status = Column(String,'RULE_EXEC_STATUS', 1011)
+    # address = Column(String,'RULE_EXEC_ADDRESS', 1004)
+    # priority = Column('RULE_EXEC_PRIORITY', 1007)
```
Comment (trel) [inline irods/models.py:35]: priority is being used in 429 now, please add it

File: irods/models.py [hunk 2]
```diff
@@ -22,22 +22,19 @@ class Model(six.with_metaclass(ModelBase, object)):
 class RuleExec(Model):
     id = Column(Integer, 'RULE_EXEC_ID', 1000)
     name = Column(String, 'RULE_EXEC_NAME', 1001)
+    rei_file_path = Column(String,'RULE_EXEC_REI_FILE_PATH', 1002)
     user_name = Column(Integer, 'RULE_EXEC_USER_NAME', 1003)
+    exec_time = Column(DateTime,'RULE_EXEC_TIME',    1005)
```
Comment (trel) [inline irods/models.py:27]: consider... just making this 'time'?

File: irods/rule.py [hunk 3]
```diff
@@ -21,6 +29,17 @@ def __init__(self, session, rule_file=None, body='', params=None, output=''):
         if output != '':
             self.output = output
 
+    #define RULE_EXEC_DEL_INP_PI "str ruleExecId[NAME_LEN];"
```
Comment (trel) [inline irods/rule.py:32]: unneeded?

File: irods/models.py [hunk 4]
```diff
@@ -22,22 +22,19 @@ class Model(six.with_metaclass(ModelBase, object)):
...
-#('RULE_EXEC_FREQUENCY', 1006)
-#('RULE_EXEC_PRIORITY', 1007)
-#('RULE_EXEC_ESTIMATED_EXE_TIME', 1008)
-#('RULE_EXEC_NOTIFICATION_ADDR', 1009)
-#(DateTime,'RULE_EXEC_LAST_EXE_TIME', 1010)
-#('RULE_EXEC_STATUS', 1011)
-#('RULE_EXEC_CONTEXT', 1012)
+# - Unused, or usually absent:
+
+    # exec_status = Column(String,'RULE_EXEC_STATUS', 1011)
```
Comment (trel) [inline irods/models.py:33]: this seems... useful.    in case a rule has failed for some reason?

Comment (d-w-moore) [reply irods/models.py:33]: Item for discuss, as this always seems to be blank.

File: irods/rule.py (hunk 3 above)
Comment (d-w-moore) [reply irods/rule.py:32]: Not needed, but (often in precedent) provided for reference.

File: irods/models.py (hunk 2 above)
Comment (d-w-moore) [reply irods/models.py:27]: If so, should  `last_exe_time` also be changed to `last`?

File: irods/models.py (hunk 4 above)
Comment (d-w-moore) [reply irods/models.py:33]: Will see what appears there for a failed rule. Seems like a success (so far) just leaves it blank? @alanking ?

Comment (trel) [reply irods/models.py:33]: yes, i think so - and i just made a bad one... and status didn't change.   so yes, let's leave it out for now.   can always add it easily later.

Comment (alanking) [reply irods/models.py:33]: `RULE_EXEC_STATUS` was used in the past to indicate whether a rule was ready to be run, currently running, failed, etc. It is no longer used due to performance and the new design for the delay server doesn't require its use (yet).

File: irods/models.py (hunk 2 above)
Comment (trel) [reply irods/models.py:27]: i don't think so...

- 'time' is the next scheduled time
- 'last_exe_time' is the last time it fired

could have an overload/convenience accessor where:
- 'next' -> 'time'
- 'previous' -> 'last_exe_time'

but then we're making up new words...

File: irods/models.py (hunk 4 above)
Comment (trel) [reply irods/models.py:33]: what he said.

Comment (d-w-moore) [reply irods/models.py:33]: Ah, never mind : ) ... haven't actually looked at that field. Also need to figure a way to include it conditionally based on server version.

Comment (trel) [reply irods/models.py:33]: no, don't worry about it.   to the future!!!

**Comment by d-w-moore** [comment]
Yes, done. ready to # if no outstanding review requests.

**Comment by trel** [comment]
yes, please squash and # it.

**Comment by d-w-moore** [comment]
Done!

**Pull request merged**